File: test/load_test.py
#!/usr/bin/env python3
"""
load_test.py — ทดสอบระบบ 30 แจ้งซ่อม + 30 จองห้อง พร้อมกัน
วิธีรัน: python3 test/load_test.py
"""

import json
import time
import urllib.request
import urllib.error
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime, timedelta

GAS_URL = (
    "https://script.google.com/macros/s/"
    "AKfycbw6joijiaVLYTOEG8ajOl7wp5_JhsovC_yW3oqJ3XWHuKvlxmxHGr5bmTY33mnhtEN1/exec"
)

ROOMS = [
    "หอประชุมซาวีโอ",
    "ห้องประชุมอัลเบรา",
    "ห้องประชุมรีกัลโดเน",
    "Auditorium",
]
EQUIPMENT = ["ไมโครโฟน", "ลำโพง / เครื่องเสียง", "โปรเจกเตอร์", "จอภาพ / จอ LED"]


# ─── HTTP POST (รองรับ redirect 302 ด้วยตนเอง) ─────────────────
def post_gas(payload: dict) -> dict:
    start = time.time()
    body = json.dumps(payload, ensure_ascii=False).encode("utf-8")

    def do_post(url):
        req = urllib.request.Request(
            url,
            data=body,
            headers={"Content-Type": "text/plain; charset=utf-8"},
            method="POST",
        )
        return urllib.request.urlopen(req, timeout=60)

    try:
        # ส่ง POST แรก — อาจได้ redirect
        try:
            resp = do_post(GAS_URL)
        except urllib.error.HTTPError as e:
            if e.code in (301, 302, 303, 307, 308):
                redirect_url = e.headers.get("Location")
                resp = do_post(redirect_url)
            else:
                raise

        elapsed = round((time.time() - start) * 1000)
        text = resp.read().decode("utf-8")
        data = json.loads(text)
        return {
            "ok": data.get("success") is True,
            "ticket": data.get("ticket"),
            "error": data.get("error"),
            "elapsed": elapsed,
        }

    except Exception as exc:
        return {
            "ok": False,
            "ticket": None,
            "error": str(exc)[:120],
            "elapsed": round((time.time() - start) * 1000),
        }


# ─── PAYLOAD FACTORIES ─────────────────────────────────────────
def make_repair(i: int) -> dict:
    return {
        "type": "repair",
        "name": f"ผู้ทดสอบ คนที่ {i + 1}",
        "phone": f"08{str(i).zfill(8)}",
        "equipment": EQUIPMENT[i % len(EQUIPMENT)],
        "location": f"อาคาร SAVIO ชั้น {(i % 4) + 1} ห้อง {100 + i}",
        "description": f"[LOAD TEST #{i + 1}] ทดสอบระบบอัตโนมัติ",
        "timestamp": datetime.now().isoformat(),
    }


def make_booking(i: int) -> dict:
    future = datetime.now() + timedelta(days=7 + i)   # กระจายวันไม่ให้ conflict
    date_str = future.strftime("%Y-%m-%d")
    start_h = 8 + (i % 10)
    return {
        "type": "booking",
        "name": f"ผู้ทดสอบ คนที่ {i + 1}",
        "phone": f"09{str(i).zfill(8)}",
        "room": ROOMS[i % len(ROOMS)],
        "date": date_str,
        "startTime": f"{start_h:02d}:00",
        "endTime":   f"{start_h + 1:02d}:00",
        "attendees": str(10 + i),
        "purpose": f"[LOAD TEST #{i + 1}] ทดสอบการจองห้องพร้อมกัน",
        "equipment": EQUIPMENT[i % len(EQUIPMENT)],
        "note": f"Auto test entry {i + 1}",
        "timestamp": datetime.now().isoformat(),
    }


# ─── HELPERS ───────────────────────────────────────────────────
def bar(ok, total, width=20):
    filled = round(ok / total * width)
    return f"[{'█' * filled}{'░' * (width - filled)}] {ok}/{total}"


def summarize(label: str, results: list[dict]):
    ok      = sum(1 for r in results if r["ok"])
    fail    = len(results) - ok
    times   = [r["elapsed"] for r in results]
    avg     = round(sum(times) / len(times))
    mn, mx  = min(times), max(times)
    tickets = [r["ticket"] for r in results if r["ticket"]]
    errors  = list({r["error"] for r in results if not r["ok"]})

    print(f"\n{'─' * 58}")
    print(f"  {label}")
    print(f"{'─' * 58}")
    print(f"  สำเร็จ    {bar(ok, len(results))}")
    print(f"  ล้มเหลว   {fail} รายการ")
    print(f"  เวลาเฉลี่ย {avg} ms  (min {mn} ms / max {mx} ms)")
    if tickets:
        sample = ", ".join(tickets[:5])
        extra  = f" … (+{len(tickets) - 5})" if len(tickets) > 5 else ""
        print(f"  Tickets   {sample}{extra}")
    for err in errors[:3]:
        print(f"  ⚠ {err}")


# ─── MAIN ──────────────────────────────────────────────────────
def main():
    N = 30
    print(f"\n{'═' * 58}")
    print(f"  🔬 LOAD TEST — {N} แจ้งซ่อม + {N} จองห้อง พร้อมกัน")
    print(f"  GAS: …{GAS_URL[-40:]}")
    print(f"{'═' * 58}\n")

    repairs  = [make_repair(i)  for i in range(N)]
    bookings = [make_booking(i) for i in range(N)]
    all_payloads = repairs + bookings        # รวม 60 รายการ

    print(f"  ⏳ กำลังส่ง {len(all_payloads)} requests (workers=20)…")
    wall_start = time.time()

    results = [None] * len(all_payloads)
    with ThreadPoolExecutor(max_workers=20) as ex:
        futures = {ex.submit(post_gas, p): i for i, p in enumerate(all_payloads)}
        done = 0
        for fut in as_completed(futures):
            idx = futures[fut]
            results[idx] = fut.result()
            done += 1
            print(f"\r  ส่งแล้ว {done}/{len(all_payloads)}…", end="", flush=True)

    wall_ms = round((time.time() - wall_start) * 1000)
    print(f"\r  ✅ เสร็จสิ้นภายใน {wall_ms} ms ({wall_ms / 1000:.1f} วินาที)       ")

    repair_res  = results[:N]
    booking_res = results[N:]

    summarize("📋 ผลแจ้งซ่อม (30 รายการ)",       repair_res)
    summarize("📅 ผลจองห้องประชุม (30 รายการ)",   booking_res)

    all_ok   = sum(1 for r in results if r["ok"])
    all_fail = len(results) - all_ok

    print(f"\n{'═' * 58}")
    print(f"  รวม: สำเร็จ {all_ok}/60  |  ล้มเหลว {all_fail}/60")
    print(f"  เวลารวม (wall clock): {wall_ms} ms")
    if all_fail == 0:
        print("  🎉 ระบบรองรับ 60 requests พร้อมกันได้ปกติ")
    else:
        print("  ⚠️  มีบางรายการล้มเหลว — ดู error ด้านบน")
    print(f"{'═' * 58}\n")


if __name__ == "__main__":
    main()

File: test/load_test.mjs
// ============================================================
// load_test.mjs — ทดสอบระบบ 30 แจ้งซ่อม + 30 จองห้อง พร้อมกัน
// วิธีรัน: node test/load_test.mjs
// ============================================================

const GAS_URL = 'https://script.google.com/macros/s/AKfycbw6joijiaVLYTOEG8ajOl7wp5_JhsovC_yW3oqJ3XWHuKvlxmxHGr5bmTY33mnhtEN1/exec';

const ROOMS = ['หอประชุมซาวีโอ', 'ห้องประชุมอัลเบรา', 'ห้องประชุมรีกัลโดเน', 'Auditorium'];
const EQUIPMENT_OPTIONS = ['ไมโครโฟน', 'ลำโพง / เครื่องเสียง', 'โปรเจกเตอร์', 'จอภาพ / จอ LED'];

// ส่ง POST ไปยัง GAS (จัดการ redirect 302 ด้วยตนเอง)
async function postGAS(payload) {
  const start = Date.now();
  try {
    const res = await fetch(GAS_URL, {
      method: 'POST',
      headers: { 'Content-Type': 'text/plain' },
      body: JSON.stringify(payload),
      redirect: 'follow',
    });

    const elapsed = Date.now() - start;

    if (!res.ok) {
      return { ok: false, status: res.status, elapsed, error: `HTTP ${res.status}` };
    }

    const text = await res.text();
    let json;
    try { json = JSON.parse(text); } catch {
      return { ok: false, elapsed, error: `Non-JSON response: ${text.slice(0, 80)}` };
    }

    return { ok: json.success === true, elapsed, ticket: json.ticket, error: json.error };
  } catch (err) {
    return { ok: false, elapsed: Date.now() - start, error: err.message };
  }
}

// สร้าง payload แจ้งซ่อม
function makeRepair(i) {
  return {
    type: 'repair',
    name: `ผู้ทดสอบ คนที่ ${i + 1}`,
    phone: `08${String(i).padStart(8, '0')}`,
    equipment: EQUIPMENT_OPTIONS[i % EQUIPMENT_OPTIONS.length],
    location: `อาคาร SAVIO ชั้น ${(i % 4) + 1} ห้อง ${100 + i}`,
    description: `[LOAD TEST #${i + 1}] ทดสอบระบบอัตโนมัติ — อุปกรณ์ทำงานผิดปกติ`,
    timestamp: new Date().toISOString(),
  };
}

// สร้าง payload จองห้อง
function makeBooking(i) {
  const today = new Date();
  const futureDate = new Date(today);
  futureDate.setDate(today.getDate() + 7 + i); // กระจายวันที่ไม่ให้ conflict
  const dateStr = futureDate.toISOString().split('T')[0];
  const startH = String(8 + (i % 8)).padStart(2, '0');
  return {
    type: 'booking',
    name: `ผู้ทดสอบ คนที่ ${i + 1}`,
    phone: `09${String(i).padStart(8, '0')}`,
    room: ROOMS[i % ROOMS.length],
    date: dateStr,
    startTime: `${startH}:00`,
    endTime:   `${String(parseInt(startH) + 1).padStart(2, '0')}:00`,
    attendees: String(10 + i),
    purpose: `[LOAD TEST #${i + 1}] ทดสอบการจองห้องพร้อมกัน`,
    equipment: EQUIPMENT_OPTIONS[i % EQUIPMENT_OPTIONS.length],
    note: `Auto test entry ${i + 1}`,
    timestamp: new Date().toISOString(),
  };
}

function printBar(ok, total) {
  const filled = Math.round((ok / total) * 20);
  return `[${'█'.repeat(filled)}${'░'.repeat(20 - filled)}] ${ok}/${total}`;
}

function summarize(label, results) {
  const ok      = results.filter(r => r.ok).length;
  const fail    = results.filter(r => !r.ok).length;
  const times   = results.map(r => r.elapsed);
  const avg     = Math.round(times.reduce((a, b) => a + b, 0) / times.length);
  const min     = Math.min(...times);
  const max     = Math.max(...times);
  const tickets = results.filter(r => r.ticket).map(r => r.ticket);
  const errors  = results.filter(r => !r.ok).map(r => r.error);

  console.log(`\n${'─'.repeat(56)}`);
  console.log(`  ${label}`);
  console.log(`${'─'.repeat(56)}`);
  console.log(`  สำเร็จ    ${printBar(ok, results.length)}`);
  console.log(`  ล้มเหลว   ${fail} รายการ`);
  console.log(`  เวลาเฉลี่ย ${avg} ms  (min ${min} ms / max ${max} ms)`);
  if (tickets.length) console.log(`  Tickets   ${tickets.slice(0, 5).join(', ')}${tickets.length > 5 ? ` … (+${tickets.length - 5})` : ''}`);
  if (errors.length) {
    console.log(`  ข้อผิดพลาด:`);
    [...new Set(errors)].slice(0, 3).forEach(e => console.log(`    • ${e}`));
  }
}

// ─── MAIN ────────────────────────────────────────────────────
async function main() {
  const N = 30;
  console.log(`\n${'═'.repeat(56)}`);
  console.log(`  🔬 LOAD TEST — ${N} แจ้งซ่อม + ${N} จองห้อง พร้อมกัน`);
  console.log(`  GAS: ${GAS_URL.slice(0, 60)}…`);
  console.log(`${'═'.repeat(56)}\n`);

  // สร้าง payloads
  const repairs  = Array.from({ length: N }, (_, i) => makeRepair(i));
  const bookings = Array.from({ length: N }, (_, i) => makeBooking(i));

  console.log(`  ⏳ กำลังส่ง ${N * 2} requests พร้อมกัน...`);
  const wallStart = Date.now();

  // ส่งทั้งหมดพร้อมกัน
  const [repairResults, bookingResults] = await Promise.all([
    Promise.all(repairs.map(p => postGAS(p))),
    Promise.all(bookings.map(p => postGAS(p))),
  ]);

  const wallTime = Date.now() - wallStart;
  console.log(`  ✅ เสร็จสิ้นภายใน ${wallTime} ms (${(wallTime / 1000).toFixed(1)} วินาที)\n`);

  summarize('📋 ผลแจ้งซ่อม (30 รายการ)', repairResults);
  summarize('📅 ผลจองห้องประชุม (30 รายการ)', bookingResults);

  const allOk = [...repairResults, ...bookingResults].filter(r => r.ok).length;
  const allFail = N * 2 - allOk;
  console.log(`\n${'═'.repeat(56)}`);
  console.log(`  รวม: สำเร็จ ${allOk}/60  |  ล้มเหลว ${allFail}/60`);
  console.log(`  เวลารวม: ${wallTime} ms`);
  if (allFail === 0) {
    console.log('  🎉 ระบบรองรับ 60 requests พร้อมกันได้ปกติ');
  } else {
    console.log('  ⚠️  มีบางรายการล้มเหลว — ตรวจสอบ log ด้านบน');
  }
  console.log(`${'═'.repeat(56)}\n`);
}

main().catch(console.error);
